server/embed/config.go

```diff
@@ -59,6 +60,7 @@ const (
 	DefaultWarningApplyDuration        = 100 * time.Millisecond
 	DefaultWarningUnaryRequestDuration = 300 * time.Millisecond
 	DefaultMaxRequestBytes             = 1.5 * 1024 * 1024
+	DefaultMaxConcurrentStreams        = math.MaxUint32
```

Comment:
This changes the default of at least 100 that is recommended by HTTP/2 spec. https://pkg.go.dev/golang.org/x/net/http2#Server 

Going from hundreds to MaxUint32 is pretty big difference. This opens a new possible issue of misbehaving problem, so the question is about expected user expirience:
* We leave the previous default, if user hits the limit, they will notice slow requests, learn that there is option to change it and they decide to increase it.
* We set MaxUint32 as default, misbehaving clients will cause performance issues for users possibly causing some clusters to go down. Client bug in popular application using etcd, makes this issue prominent and users figure out new safe default by themselves. 

I would prefer to separate opening new options to users from changing the defaults.  I don't think that provided examples would motivate us to change the default for everyone.

Reply:
It seems that the best choice for us is to use `math.MaxUint32` as the default value. Reasons:
1. The existing maxStream for grpc is already [math.MaxUint32](https://github.com/etcd-io/etcd/blob/248f072370628fa2ad7c3d73e3935ba5636c0442/server/etcdserver/api/v3rpc/grpc.go#L35), changing the value may breaking the existing behavior;
2. The [HTTP2 spec](https://datatracker.ietf.org/doc/html/rfc7540#section-6.5.2) says "no smaller than 100", but does not say that it can't be MaxUint32 (of course, this point may be a little debatable). 
```
      Initially, there is no limit to
      this value.  It is recommended that this value be no smaller than
      100, so as to not unnecessarily limit parallelism.
```
3. Actually the existing implementation of gRPC also set the default value as [math.MaxUint32](https://github.com/grpc/grpc-go/blob/5b509df1e30a25b14dba119c09295924c9e17abe/internal/transport/http2_server.go#L172).

Reply:
Ok, I'm getting a little confused about what bugfix vs change of default. https://github.com/etcd-io/etcd/issues/14077#issuecomment-1140057917 mentions that the current ConcurrentMaxStream for TLS enabled clusters is 250. In above comment you say to set ConcurrentMaxStream to max.MaxUint32 which is default for grpc, so why it was not used by etcd before? Are we making changes or not?

Based on that we are changing the default which would be something that I would prefer not to backport.

Can you please summarize:
* What this PR changes?
* Why do you think it's worth backporting this?

Reply:
They are indeed good questions. The points are:

1. The default maxStream for gRPC is [math.MaxUint32](https://github.com/grpc/grpc-go/blob/5b509df1e30a25b14dba119c09295924c9e17abe/internal/transport/http2_server.go#L172), while the default value for http2 is [250](https://cs.opensource.google/go/x/net/+/master:http2/server.go;drc=1d687d428aca0546c0ca84160c8700ee521e9fb9;l=58).

2. Why does only https have such issue? If you have this question as well, please see https://github.com/etcd-io/etcd/pull/13985#issuecomment-1120047735 

3. Actually previous PR [8238](https://github.com/etcd-io/etcd/pull/8238) isn't complete, because it only changed the `maxStream` for gRPC, but not for HTTP2. But probably the [code of HTTP encapsulation gRPC](https://github.com/etcd-io/etcd/blob/248f072370628fa2ad7c3d73e3935ba5636c0442/server/embed/serve.go#L126-L207) might be added after the PR, I do not waste my time to double check that; if it's the case, then we can say the change of encapsulating gRPC using HTTP has issue.  Anyway, no matter which case is true, the issue is the max concurrent stream each client can open at a time is actually 250 instead of math.MaxUint32 over https.  It's an issue, so we need to resolve it, and accordingly backport to 3.5. 

Do you still have any concern?

Reply:
Regarding how to the cherry pick the PR to 3.5, it's open to discussion. Probably we can just set the maxStream as `math.MaxUint32` for http2, and do not add any command line parameters for 3.5. What do you think?

Reply:
Probably @liggitt doesn't have the full context, so let me clarify. Previously in PR [8238](https://github.com/etcd-io/etcd/pull/8238), the `grpc.MaxConcurrentStreams` was set to `math.MaxUint32` by default. But the PR did not change the [MaxConcurrentStreams for http2](https://cs.opensource.google/go/x/net/+/master:http2/server.go;l=99;drc=1d687d428aca0546c0ca84160c8700ee521e9fb9;bpv=0;bpt=1).

etcd [encapsulates the gRPC in https](https://github.com/etcd-io/etcd/blob/248f072370628fa2ad7c3d73e3935ba5636c0442/server/embed/serve.go#L166-L172), which means each gRPC request goes through httpHandler --> gRPC. Sicne the [default MaxConcurrenetStreams is 250](https://cs.opensource.google/go/x/net/+/master:http2/server.go;drc=1d687d428aca0546c0ca84160c8700ee521e9fb9;l=58) for http2, so actually the max concurrent stream each client can open at a time is actually 250 instead of math.MaxUint32 over https. 

Please let us know if you have any comments or concerns.  thx. also cc @dims

Reply:
Lgtm